### DISCUSSION TITLE
ImageGlass Moon 9.0 beta 4 🌔

### DISCUSSION THREAD

**Discussion by d2phap** [d2phap/ImageGlass#1541] [General]
<img width="640" alt="ImageGlass 9.0 beta 4" src="https://github.com/ImageGlass/website-content/blob/main/releases/screenshots/v9.0-beta-4/9.0b4_2.jpg?raw=true">

[![Downloads](https://img.shields.io/github/downloads/d2phap/imageglass/9.0.4.416/total?color=%232A7C91&label=downloads&style=for-the-badge)](https://github.com/d2phap/ImageGlass/releases/download/9.0.4.416/ImageGlass_Moon_9.0.4.416_Beta-4_x64.zip)

## ⚠ Notes
This is the beta-4 release of ImageGlass version 9. Features and settings may still be changed when the official version 9 is released.

To follow the progress of version 9 beta 5, please visit: https://github.com/d2phap/ImageGlass/projects/15.
You also can support this project by making a donation or purchasing the Microsoft Store version (version 8):

<a href="https://www.microsoft.com/store/productId/9N33VZK3C7TH?cid=GitHubRelease&referrer=appbadge&source=GithubD2phap">
  <img height="69" src="https://user-images.githubusercontent.com/3154213/197671083-4099d80f-07fc-4af9-b80d-6357f0968347.svg" />
</a>

## 🎐 Minimum requirements
- [.NET Desktop Runtime 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0)
- 🆕 [WebView2 Runtime x64](https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section)
- 🆕 Windows 10 64-bit

# 📝 Changelog:
ImageGlass 9.0 beta 4 is the continuation of the version 9.0 beta 3,
you should read the previous changelog here: https://github.com/d2phap/ImageGlass/releases/tag/9.0.3.1225

<details>
<summary>

### Show / hide details

</summary>

## 🟣 General changes
- Upgraded Magick.NET to [version 13.0.1](https://github.com/dlemstra/Magick.NET/releases/tag/13.0.1).
- Improved ImageGlass startup time and latency by enabling [ReadyToRun compilation](https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run).
- Showed Color space and Color profile info on the title bar.
- Added basic support for animated APNG format (#396).
- Fixed a bug where ImageGlass cannot load image with CMYK color space. Error message: "_Operation is not valid due to the state of the object_".
- Fixed a bug that caused memory leak during loading image.
- Skipped file existence check while loading image files into the list (#585).
- Changed the prefix for passing app setting as command-line from `-` to `/`. Example: `ImageGlass.exe /EnableFrameless=true`.

## 🟣 Theme component
- Added support for auto-switching theme according to system's dark/light mode (#434).
  + Splits the setting `Theme` into `DarkTheme` and `LightTheme`.
  + Loads `LightTheme` if system light mode is on, otherwise ImageGlass uses `DarkTheme`.
- Removed accent color in theme config.
- Added new icons for toolbar: `Play`, `Pause`, `Export`.

## 🟣 Viewer component
- Improved frame animation (GIF, WEBP, ...)
  + Adds support for hardware acceleration.
  + Supports playing, pausing or viewing a single frame of image.
  + Adds basic support for playing animation for all animated formats using MagickNET, such as APNG.
- Made the navigation arrow buttons clickable on whole left and right areas (#1472).

## 🟣 Thumbnail bar component
- Shows Color space and Color profile info on the thumbnail's tooltip.

## 🟣 Image Booster component
- Reduced memory usage when loading non-transparent image.
- Enabled basic support for all animated formats using MagickNET.
- Cached metadata to improve loading image.

## 🟣 Features
- Added support for updating the viewing image folder in real time:
  + Updates the viewing image, thumbnail when it's changed by third-party app.
  + Updates the thumbnail bar when image file is deleted, added, renamed.
  + Added an option to disable it by setting: `EnableFileWatcher`.
- Added Edit image with external editor:
  + Added setting `EditApps`, example:
    ```js
    // in igconfig.json
    "EditApps": {
      ".png": {
        "AppName": "Paint.NET",
        "Executable": "C:\\paint.net\\paintdotnet.exe",
        "Argument": "<file>"
      }
    },
    ```
  + Added options what to do after the external editor is launch `AfterEditingAction`:
    - `Nothing`: Do nothing (default).
    - `Minimize`: Minimize ImageGlass window.
    - `Close`: Close ImageGlass app.
- Added feature: Export image frames
  + Moves the function to `igcmd.exe` with command `export-frames`.
  + Shows exporting progress and status.
  + Added a button to cancel the on-going export progress.
  + Added a button to open output folder after the export is done.
- Added feature: Rotate image (only for non-animated format).
- Added image transformation features:
  + Supports image rotation.
  + Improves performance of image flipping / ratation by using hardware acceleration.
  + Shows asterisk `*` on title bar next to the image name to indicate there is unsaved change.
  + Displays "Save..." / "Save as..." button on toolbar if the image is changed.
- Adds native support for animated Webp format (#1466) using `libwebp.dll` v1.3.0:
  + Supports hardware acceleration.
  + Fixed: Quality of animated Webp is reduced (#1526).
  + Supports saving image as Webp, including animated format (#1477).
- Supports for saving TIFF, GIF as multiple frames.
- Improves ImageGlass Slideshow view:
  + Removed support for multiple slideshow views
  + Added support for changing slideshow image list when user changes it in the main view.
  + Made slideshow view togglable.
  + Added support for new window modes in slideshow view.

### 🟢 New window modes
- Frameless mode
  + The window corners are rounded by default.
  + Hold SHIFT and drag to move the window.
- Window fit mode
  + Disables Maximize button of the window.
  + Improve performance when user zooms the image.
  + Added support for all zoom modes in Window mode.
- Window modes combination for the main view and slideshow view:
  | Mode | Full screen | Frameless | Window fit |
  | -- | -- | -- | -- |
  | Full screen | ✅ | ❌ | ❌ |
  | Frameless | ❌ | ✅ | ✅ |
  | Window fit | ❌ | ✅ | ✅ |

### 🟢 Layout customization
- Added support for layout customization using setting `Layout` (#318) for:
  + Toolbar with setting `Toolbar`,
  + Thumbnail bar with setting `Gallery`,
  + Contextual toolbar with setting `ToolbarContext`.
- Supports Toolbar position: `Top`, `Bottom`.
- Supports Thumbnail bar position: `Top`, `Bottom`, `Left`, `Right`.
- Allows to change control's order if they are in the same layout position.
- Example:
  + Default layout:
    ```js
    // in igconfig.json
    {
      // default layout
      "Layout": {
        "Toolbar": "Top", // Toolbar is at top
        "Gallery": "Bottom;0", // thumbnail bar is at bottom
        "ToolbarContext": "Bottom;1", // contextual toolbar is at bottom, above thumbnail bar
      },
    }
    ```
  + Layout: toolbar and thumbnail bar at top, contextual toolbar at bottom
    ```js
    // in igconfig.json
    {
      "Layout": {
        "Toolbar": "Top;1", // Toolbar is at top, below thumbnail bar
        "Gallery": "Top;0", // thumbnail bar is at top, above the toolbar
        "ToolbarContext": "Bottom", // contextual toolbar is at bottom
      }
    }
    ```

### 🟢 New Color picker tool

- Shows cursor location on the viewing image.
- Clicks to pick the pixel color.
- Adds support for color formats: RGB, HEX, CMYK, HSL, HSV.
- Adds buttons to copy color code.
- Adds Settings window with options to show alpha value for each color format.

### 🟢 New Page navigation tool
- Supports for viewing separate image frame.
  + Non-animated formats (TIFF, ICO,...).
  + Animated formats (GIF, WEBP, APNG,...).
- Respects the size and orientation of the image frame (#1077).
- Exports all frames.

### 🟢 External tools support
- You can download external tools at https://imageglass.org/tools.
- Created https://github.com/ImageGlass/ImageGlass.Tools library for other 3rd-party apps to communicate with ImageGlass via IPC message events.
- Sends events to the external tools using `ImageGlassEvents`.
- Sends `EnableWindowTopMost` as a command-line argument to request the external tool window to be top most.
- Uses `ToolSettings` setting to store settings of the built-in tools (Crop tool, Color picker) instead of `Tools`.
- Added support for adding external tools to ImageGlass Tools menu with setting `Tools` and assigning the tool's hotkeys with setting `MenuHotkeys`.
- Shows "Tool not found" dialog to allow user to select executable path of the tool.
- For more details, please refer to https://github.com/ImageGlass/ImageGlass.Tools.

</details>

### 🟢 Introducing `ExifGlass` - a new ExifTool for ImageGlass
ExifGlass is a brand new EXIF metadata viewing app, designed to work seamlessly with ImageGlass 9, but can also be used as a standalone software on your computer. To use ExifGlass, you need to have _ExifTool by Phil Harvey_ installed on your system.

For more details, please visit: https://github.com/d2phap/ExifGlass.

[![ExifGlass](https://github.com/ImageGlass/website-content/raw/main/releases/screenshots/v9.0-beta-4/9.0b4_3.jpg?raw=true)](https://imageglass.org/tools)

## 🥒 Checksums:

File | SHA-1
-- | --
ImageGlass_Moon_9.0.4.416_Beta-4_x64_net6.zip | `CCE6542B9952779972A50FF350984A6961A097D2`
ImageGlass_Moon_9.0.4.416_Beta-4_x64_net7.zip | `1EFC8419A4364709A9EDAF60E51A2CC230DDCF9E`

<hr /><em>This discussion was created from the release <a href='https://github.com/d2phap/ImageGlass/releases/tag/9.0.4.416'>ImageGlass Moon 9.0 beta 4 🌔</a>.</em>

**Comment by VitalSkib** [comment]
Hi. 
I apologize in advance if I ask my question not in the appropriate place.. 
Is there any way to replace the `MS Paint` in the `Edit Image` menu to Photoshop? 
I can see how it is implemented in the igconfig.xml file (from version 8.x.x.x), but I have no idea how to convert it into the .json format. 
Any clues?
tia

![2023_04_14](https://user-images.githubusercontent.com/43848892/232117404-e9de7625-01d8-4b4c-857a-eacaa78e3110.png)

**Comment by VitalSkib** [comment]
Oh, I was able to deal with it, sorry..
```
"EditApps": {
    ".png": {
      "AppName": "Photoshop",
      "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
      "Argument": "<file>"
    }
  },
```
Now I have to figure out how to add other extensions, because for now it works only for .png.
I would be glad to see any tips.. 

![2023_04_14](https://user-images.githubusercontent.com/43848892/232135042-507139b6-ddc4-4e5d-8401-706333d3ab1d.png)

**Reply by d2phap** [reply]
You can do the same for other extensions:

```js
  "EditApps": {
    ".png": {
      "AppName": "Photoshop",
      "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
      "Argument": "<file>"
    },
    
    ".jpg": {
      "AppName": "Photoshop",
      "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
      "Argument": "<file>"
    },
    
    ".psd": {
      "AppName": "Photoshop",
      "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
      "Argument": "<file>"
    },
    
    ".tif": {
      "AppName": "Photoshop",
      "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
      "Argument": "<file>"
    }
  },
```

**Reply by iamLazyCode** [reply]
> You can do the same for other extensions:
> 
> ```js
>   "EditApps": {
>     ".png": {
>       "AppName": "Photoshop",
>       "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
>       "Argument": "<file>"
>     },
>     
>     ".jpg": {
>       "AppName": "Photoshop",
>       "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
>       "Argument": "<file>"
>     },
>     
>     ".psd": {
>       "AppName": "Photoshop",
>       "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
>       "Argument": "<file>"
>     },
>     
>     ".tif": {
>       "AppName": "Photoshop",
>       "Executable": "C:\\Program Files\\Adobe\\Adobe Photoshop 2023\\photoshop.exe",
>       "Argument": "<file>"
>     }
>   },
> ```

hello sir i love the app but kinda new to this can you recommend photo editor replacement i am not using paint , also where is this option to edit file ?

**Comment by VitalSkib** [comment]
Yes, it works, thank you very much.. 👍
I tried different variations, including this one, but obviously I did something wrong and it didn't worked for me, resetting the config to default.
But now everything is fine.
